src/jit/lowerxarch.cpp

```diff
@@ -2576,6 +2576,17 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
     GenTree* op2 = node->gtGetOp2();
     GenTree* op3 = nullptr;
 
+    if (HWIntrinsicInfo::lookupCategory(intrinsicId) == HW_Category_IMM)
+    {
+        GenTree* lastOp = HWIntrinsicInfo::lookupLastOp(node);
+        assert(lastOp != nullptr);
+
+        if (HWIntrinsicInfo::isImmOp(intrinsicId, lastOp) && lastOp->IsCnsIntOrI())
+        {
+            MakeSrcContained(node, lastOp);
+        }
+    }
+
     if (!HWIntrinsicInfo::SupportsContainment(intrinsicId))
     {
         // Exit early if containment isn't supported
```

Comment:
You're no longer really exiting early - was there some reason you moved the immediate containment above this?

Reply:
Before this PR, IMM intrinsics do not have NoContainment ones. But AVX2 `Gather*` are IMM intrinsic AND memory-load intrinsic that cannot work with containment (so they have NoContainment flag).

So we need to moving the immediate containment above the early exiting, which marks the IMM argument of Gather before return.

Reply:
Before this PR, IMM intrinsics do not have NoContainment ones. But AVX2 `Gather*` are IMM intrinsic AND memory-load intrinsic that cannot work with containment (so they have NoContainment flag).

So we need to moving the immediate containment above the early exiting, which marks the IMM argument of Gather before return.

Reply:
So then I think this comment needs to be deleted.